src/pk/rsa/rsa_import_pkcs8.c

```diff
@@ -0,0 +1,151 @@
+ *
+ * PrivateKeyInfo ::= SEQUENCE {
+ *      version                   Version,
+ *      privateKeyAlgorithm       PrivateKeyAlgorithmIdentifier,
+ *      privateKey                PrivateKey,
+ *      attributes           [0]  IMPLICIT Attributes OPTIONAL }
+ * where:
+ * - Version ::= INTEGER
+ * - PrivateKeyAlgorithmIdentifier ::= AlgorithmIdentifier
+ * - PrivateKey ::= OCTET STRING
+ * - Attributes ::= SET OF Attribute
+ *
+ * EncryptedPrivateKeyInfo ::= SEQUENCE {
+ *        encryptionAlgorithm  EncryptionAlgorithmIdentifier,
+ *        encryptedData        EncryptedData }
+ * where:
+ * - EncryptionAlgorithmIdentifier ::= AlgorithmIdentifier
+ * - EncryptedData ::= OCTET STRING
+ */
+
+/**
+  Import an RSAPublicKey or RSAPrivateKey in PKCS#8 format
+  @param in        The packet to import from
+  @param inlen     It's length (octets)
+  @param passwd    The password for decrypting privkey (NOT SUPPORTED YET)
+  @param passwdlen Password's length (octets)
+  @param key       [out] Destination for newly imported key
+  @return CRYPT_OK if successful, upon error allocated memory is freed
+*/
+int rsa_import_pkcs8(const unsigned char *in, unsigned long inlen,
+                     const void *passwd, unsigned long passwdlen,
+                     rsa_key *key)
+{
+   int           err;
+   void          *zero, *iter;
+   unsigned char *buf1 = NULL, *buf2 = NULL;
+   unsigned long buf1len, buf2len;
+   unsigned long oid[16];
+   oid_st        rsaoid;
+   ltc_asn1_list alg_seq[2], top_seq[3];
+   ltc_asn1_list alg_seq_e[2], key_seq_e[2], top_seq_e[2];
+   unsigned char *decrypted = NULL;
+   unsigned long decryptedlen;
+
+   LTC_ARGCHK(in          != NULL);
+   LTC_ARGCHK(key         != NULL);
+   LTC_ARGCHK(ltc_mp.name != NULL);
+
+   /* get RSA alg oid */
+   err = pk_get_oid(PKA_RSA, &rsaoid);
+   if (err != CRYPT_OK) { goto LBL_NOFREE; }
+
+   /* alloc buffers */
+   buf1len = inlen; /* approx. */
+   buf1 = XMALLOC(buf1len);
+   if (buf1 == NULL) { err = CRYPT_MEM; goto LBL_NOCLEAR; }
+   buf2len = inlen; /* approx. */
+   buf2 = XMALLOC(buf2len);
+   if (buf2 == NULL) { err = CRYPT_MEM; goto LBL_FREE; }
+
+   /* init key */
+   err = mp_init_multi(&key->e, &key->d, &key->N, &key->dQ, &key->dP, &key->qP, &key->p, &key->q, &zero, &iter, NULL);
+   if (err != CRYPT_OK) { goto LBL_NOCLEAR; }
+
+   /* try to decode encrypted priv key */
+   LTC_SET_ASN1(key_seq_e, 0, LTC_ASN1_OCTET_STRING, buf1, buf1len);
+   LTC_SET_ASN1(key_seq_e, 1, LTC_ASN1_INTEGER, iter, 1UL);
+   LTC_SET_ASN1(alg_seq_e, 0, LTC_ASN1_OBJECT_IDENTIFIER, oid, 16UL);
+   LTC_SET_ASN1(alg_seq_e, 1, LTC_ASN1_SEQUENCE, key_seq_e, 2UL);
+   LTC_SET_ASN1(top_seq_e, 0, LTC_ASN1_SEQUENCE, alg_seq_e, 2UL);
+   LTC_SET_ASN1(top_seq_e, 1, LTC_ASN1_OCTET_STRING, buf2, buf2len);
+   err=der_decode_sequence(in, inlen, top_seq_e, 2UL);
+   if (err == CRYPT_OK) {
+      LTC_UNUSED_PARAM(passwd);
+      LTC_UNUSED_PARAM(passwdlen);
+      /* XXX: TODO encrypted pkcs8 not implemented yet */
```

Comment:
I just remembered that there was something on the ML https://groups.google.com/d/topic/libtom/sFdia0rRcfM/discussion

I don't know if this is what we need here, but it looks kinda similar :)

Will you have a look? Otherwise I'll do it when I have the time.